### PR TITLE
feat(wallet-tests): add peg_ins_that_are_unconfirmed_are_rejected

### PR DESCRIPTION
**Pull request by douglaz**
Fully migrate an old integration test that was removed (https://github.com/fedimint/fedimint/pull/2969#pullrequestreview-1581944324)

Update:

Found that some code was mixing block height and block count without the proper precautions. Fixed that

### CODE REVIEW

File: modules/fedimint-wallet-tests/Cargo.toml [hunk 1]
```diff
@@ -12,21 +12,23 @@ path = "tests/tests.rs"
...
 fedimint-logging = { path = "../../fedimint-logging" }
-tokio = { version = "1.26.0", features = ["sync"] }
+fedimint-server = { path = "../../fedimint-server" }
+fedimint-testing ={ path = "../../fedimint-testing" }
+fedimint-wallet-client = { path = "../fedimint-wallet-client" }
+fedimint-wallet-common = { path = "../fedimint-wallet-common" }
+fedimint-wallet-server = { path = "../fedimint-wallet-server" }
 futures = "0.3.28"
+miniscript = { version = "9.0.2" }
+tokio = { version = "1.26.0", features = ["sync"] }
```
Comment (elsirion) [inline modules/fedimint-wallet-tests/Cargo.toml:33]: Is there an easy way to automatically order dependencies alphabetically so we can enforce it in CI?

File: modules/fedimint-wallet-tests/tests/tests.rs [hunk 2]
```diff
@@ -230,3 +240,163 @@ async fn peg_outs_support_rbf() -> anyhow::Result<()> {
...
+        Err(e) => {
+            assert!(e.to_string().contains("Unknown block hash in peg-in proof"));
+        }
+    }
+
+    // Mine the rest of the blocks
+    bitcoin
+        .mine_blocks((wallet_config.consensus.finality_delay - 1).into())
+        .await;
+    let block_count = dyn_bitcoin_rpc.get_block_count().await? - 1;
```
Comment (elsirion) [inline modules/fedimint-wallet-tests/tests/tests.rs:332]: Should the variable be called `block_height` if it's `block_count - 1`?

File: modules/fedimint-wallet-tests/tests/tests.rs [hunk 3]
```diff
@@ -230,3 +240,163 @@ async fn peg_outs_support_rbf() -> anyhow::Result<()> {
...
+        0,
+        <WalletGen as fedimint_core::module::ServerModuleInit>::kind(),
+        fedimint_core::module::ModuleConsensusVersion(0),
+        fedimint_core::module::ServerModuleInit::get_client_config(
+            &WalletGen,
+            &wallet_cfg[&PeerId::from(0)].consensus,
+        )?,
+    )?;
+    Ok((wallet_cfg.into_values().collect(), client_cfg))
+}
```
Comment (elsirion) [inline modules/fedimint-wallet-tests/tests/tests.rs:402]: We can probably deduplicate parts of this function, something similar seems to be needed for each module:

https://github.com/fedimint/fedimint/blob/af41ed842d174ffdd80f5ab274594fcae36fc962/modules/fedimint-mint-server/src/lib.rs#L783-L806

**Comment by douglaz** [comment]
There is a off by one somewhere

File: modules/fedimint-wallet-tests/tests/tests.rs [hunk 4]
```diff
@@ -78,9 +93,50 @@ async fn await_consensus_to_catch_up(client: &Client, block_count: u64) -> anyho
...
         } else {
-            break;
+            info!("Current consensus block count is {current_consensus}, consensus caught up");
+            return Ok(current_consensus);
         }
     }
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn sanity_check_bitcoin_blocks() -> anyhow::Result<()> {
```
Comment (douglaz) [inline modules/fedimint-wallet-tests/tests/tests.rs:103]: Secondary test just to straight things up.

File: fedimint-testing/src/btc/real.rs [hunk 5]
```diff
@@ -56,10 +57,12 @@ impl BitcoinTest for RealBitcoinTestNoLock {
                 .client
                 .get_block_header_info(block_hash)
                 .expect("rpc failed");
+            let expected_block_count = last_mined_block.height as u64 + 1;
             // waits for the rpc client to catch up to bitcoind
-            while self.rpc.get_block_count().await.expect("rpc failed")
-                < last_mined_block.height as u64
-            {}
+            while self.rpc.get_block_count().await.expect("rpc failed") < expected_block_count {
```
Comment (douglaz) [inline fedimint-testing/src/btc/real.rs:62]: Yes, another off by one error :man_shrugging:

File: devimint/src/external.rs [hunk 6]
```diff
@@ -80,7 +80,13 @@ impl Bitcoind {
         info!(block_num, "Mining bitcoin blocks");
         let client = self.client();
         let addr = client.get_new_address(None, None)?;
+        let initial_block_count = client.get_block_count()?;
         client.generate_to_address(block_num, &addr)?;
+        while client.get_block_count()? < initial_block_count + block_num {
```
Comment (douglaz) [inline devimint/src/external.rs:85]: Not sure this is an issue, but better safe than sorry.

File: fedimint-testing/src/btc/real.rs [hunk 7]
```diff
@@ -56,10 +57,12 @@ impl BitcoinTest for RealBitcoinTestNoLock {
                 .client
                 .get_block_header_info(block_hash)
                 .expect("rpc failed");
+            let expected_block_count = last_mined_block.height as u64 + 1;
```
Comment (douglaz) [inline fedimint-testing/src/btc/real.rs:60]: Yes, another off by one error :man_shrugging:

File: modules/fedimint-wallet-tests/tests/tests.rs [hunk 8]
```diff
@@ -46,17 +58,20 @@ async fn peg_in<'a>(
...
+        .context("expected tx to be mined")?;
+    info!(?height, "Peg-in transaction mined");
     let sub = client.subscribe_deposit_updates(op).await?;
     let mut sub = sub.into_stream();
     assert_eq!(sub.ok().await?, DepositState::WaitingForTransaction);
     assert_eq!(sub.ok().await?, DepositState::WaitingForConfirmation);
 
-    // Need to mine blocks until deposit is confirmed, but send_and_mine_block
-    // already mined one, so we can mine one less here.
-    bitcoin.mine_blocks(finality_delay - 1).await;
```
[the remark this answers is not in the document]
Comment (elsirion) [reply modules/fedimint-wallet-tests/tests/tests.rs:59]: You are right :see_no_evil: but why did tests work then?! :sob:

File: fedimint-testing/src/btc/real.rs [hunk 9]
```diff
@@ -56,10 +57,12 @@ impl BitcoinTest for RealBitcoinTestNoLock {
...
             // waits for the rpc client to catch up to bitcoind
-            while self.rpc.get_block_count().await.expect("rpc failed")
-                < last_mined_block.height as u64
-            {}
+            while self.rpc.get_block_count().await.expect("rpc failed") < expected_block_count {
+                trace!("Waiting for blocks to be mined");
+                sleep(Duration::from_millis(200)).await;
+            }
```
Comment (elsirion) [inline fedimint-testing/src/btc/real.rs:65]: Is this a case for `poll`?

File: devimint/src/external.rs [hunk 10]
```diff
@@ -80,7 +80,13 @@ impl Bitcoind {
...
+            trace!(LOG_DEVIMINT, "Waiting for blocks to be mined");
+            sleep(Duration::from_millis(200)).await;
+        }
```
Comment (elsirion) [inline devimint/src/external.rs:88]: Is this a case for `poll`?

File: modules/fedimint-wallet-server/src/lib.rs [hunk 11]
```diff
@@ -393,7 +393,14 @@ impl ServerModule for Wallet {
                 // only sync when we have a consensus block count
                 match (old_consensus_block_count, new_consensus_block_count) {
                     (Some(old), Some(new)) if new > old => {
-                        self.sync_up_to_consensus_height(dbtx, old, new).await;
+                        if old > 0 {
+                            let new_height = new - 1;
+                            let old_height = old - 1;
+                            self.sync_up_to_consensus_height(dbtx, old_height, new_height)
```
[the remark this answers is not in the document]
Comment (elsirion) [reply modules/fedimint-wallet-server/src/lib.rs:399]: I'd slightly prefer also migrating `sync_up_to_consensus_height` to count instead of height.

**Review by elsirion — APPROVED** [review]
LGTM, only some nits and things that would fit better into a follow-up.

**Pull request merged**